### DISCUSSION TITLE
Loading Model With Randomly Initialized Weights

### DISCUSSION THREAD

**Discussion by brucejwittmann** [facebookresearch/esm#87] [Q&A]
Hi all, 

I was wondering if there was an inbuilt way in the ESM repo to load models with randomly initialized weights. Essentially, the goal is to train a new network using the ESM model architectures as a component. Something like the below:

```
class NewModule(nn.module):

    def __init__(self):
         super().__init__()
         self.esm_transformer = RandomlyInitialiedEsmModel
         self.custom_layers = CustomLayers

    def forward(self, x):
         x = self.esm_transformer(x)
         return self.custom_layers(x)
```

From reading through the code, it looks like this could be achieved by modifying the function `load_model_and_alphabet_core` to remove the `model.load_state_dict` at [this](https://github.com/facebookresearch/esm/blob/0097c27290c7c88c4bc0fc4df3692d2c23a6aa54/esm/pretrained.py#L114) line. Is this a legitimate way to achieve random initialization? Is there a recommended way to go about this?

**Comment by joshim5** [comment]
Hey Bruce, good question! The answer depends on your goal. If you skip that line, then the weights won't be loaded but you won't necessarily have the ideal random initialization for training.

If you want to train the model from scratch, then initialization can have a strong impact on the training and convergence of the model. In that case, I would recommend following the initializations described in our paper, which don't strictly follow the PyTorch defaults.

**Reply by brucejwittmann** [reply]
Hey Josh,

Thanks for the quick response! That makes sense. The goal is to train the model from scratch, so perhaps the easiest way to randomly initialize models while still using the inbuilt ESM functions is to manually reset the parameters after loading a pretrained model. Based on your paper, the first step of initialization for all parameters was random from a zero-centered normal distribution with a 0.02 standard deviation. So the below should handle that, right (using esm1_t6_43M_UR50S as an example)?

```
# Load model
model, alphabet = esm.pretrained.esm1_t6_43M_UR50S()

# Re-initialize all learnable parameters
for trainable_tensor in model.parameters():
    torch.nn.init.normal_(trainable_tensor.data, mean=0.0, std=0.02)
```

The next step was then to "re-scale the initialization of the projections into the residual path by 1/sqrt(L), where L is the number of residual layers." Just to make sure I understand this step, the number of residual layers will be 2x the number of transformer layers -- one residual layer output from each self-attention block and another from each feed-forward block. For example, for esm1_t6_43M_UR50S, "L" would be 12. By the "projections", do you mean rescaling the weights/biases of just the linear layer at the end of the multi-head attention block, or does this include the outputs of the feed-forward blocks as well? Assuming just the linear layer of the attention block, I believe the below code would be appropriate for rescaling its weights/biases, right? 

```

# Get the residual layer scaling factor
n_residual_layers = 12 # Will change depending on the number of transformer layers
residual_scale = 1 / math.sqrt(n_residual_layers)

# Rescale weights for layers that output to a residual layer
for layer_name, layer_tensor in model.state_dict().items():
    if "out_proj" in layer_name:
        layer_tensor *= residual_scale
```
